### PR TITLE
Fix AEAD corner cases

### PR DESCRIPTION
**Pull request by ni4**
This PR fixes found corner cases in rnp which didn't work properly.
Now it is finally 100% compatible with latest GnuPG master.
Enables AEAD CLI tests as well.

### CODE REVIEW

File: src/librepgp/stream-parse.c [hunk 1]
```diff
@@ -603,7 +616,8 @@ encrypted_src_read_aead_part(pgp_source_encrypted_param_t *param)
         }
     }
 
-    res = encrypted_start_aead_chunk(param, chunkend ? param->chunkidx + 1 : param->chunkidx, lastchunk);
+    res = encrypted_start_aead_chunk(
+      param, chunkend && param->chunkin ? param->chunkidx + 1 : param->chunkidx, lastchunk);
```
Comment (dewyatt) [inline src/librepgp/stream-parse.c]: I think some parentheses around the ternary conditional expr. would make this a bit clearer.

File: src/lib/symmetric.c [hunk 2]
```diff
@@ -613,6 +613,18 @@ pgp_cipher_aead_update(pgp_crypt_t *crypt, uint8_t *out, const uint8_t *in, size
...
+void
+pgp_cipher_aead_reset(pgp_crypt_t *crypt)
+{
+    uint32_t flags = BOTAN_CIPHER_UPDATE_FLAG_FINAL;
+    uint8_t  tag[PGP_AEAD_EAX_OCB_TAG_LEN] = {0};
+    size_t   in = 0, out = 0;
+    size_t   len = crypt->aead.decrypt ? crypt->aead.taglen : 0;
+    size_t   dlen = crypt->aead.decrypt ? 0 : crypt->aead.taglen;
+
+    botan_cipher_update(crypt->aead.obj, flags, tag, dlen, &out, tag, len, &in);
```
Comment (henrydcase) [inline src/lib/symmetric.c]: Why not to simply use "botan_cipher_clear"

File: src/librepgp/stream-write.c [hunk 3]
```diff
@@ -331,8 +331,16 @@ encrypted_start_aead_chunk(pgp_dest_encrypted_param_t *param, size_t idx, bool l
     STORE64BE(param->ad + param->adlen - 8, idx);
 
     if (last) {
-        total = idx ? (idx - 1) * param->chunklen : 0;
-        total += param->cachelen + param->chunkout;
+        if (param->chunkout + param->cachelen == 0) {
```
Comment (henrydcase) [inline src/librepgp/stream-write.c]: parenthesis also here, please

File: src/librepgp/stream-write.c [hunk 4]
```diff
@@ -331,8 +331,16 @@ encrypted_start_aead_chunk(pgp_dest_encrypted_param_t *param, size_t idx, bool l
...
+            /* we need to clearly reset it since cipher was initialized but not finished */
+            pgp_cipher_aead_reset(&param->encrypt);
+        }
+
+        total = idx * param->chunklen;
+        if (param->cachelen + param->chunkout) {
+            total -= param->chunklen - param->cachelen - param->chunkout;
```
Comment (henrydcase) [inline src/librepgp/stream-write.c]: Code is a bit complicated, but I think in theory it is possible that ``chunklen < chachelen``. Which will result in bug here

**Comment by ronaldtse** [comment]
Thanks @ni4 , this looks good once the concerns raised by @henrydcase are resolves!

File: src/librepgp/stream-parse.c (hunk 1 above)
Comment (ni4) [reply src/librepgp/stream-parse.c]: Thanks. Replaced it with additional variable and if, seems to look a bit better.

File: src/lib/symmetric.c (hunk 2 above)
Comment (ni4) [reply src/lib/symmetric.c]: It will also clear the key, while it is reused with each next chunk.

File: src/librepgp/stream-write.c (hunk 3 above)
Comment (ni4) [reply src/librepgp/stream-write.c]: Okay, thanks

File: src/librepgp/stream-write.c (hunk 4 above)
Comment (ni4) [reply src/librepgp/stream-write.c]: It is safe to use since in the calling function it is checked.
However I also added check here, 'just in case'.

**Comment by ni4** [comment]
@henrydcase @dewyatt Please see the updated PR.

File: src/lib/symmetric.c (hunk 2 above)
Comment (henrydcase) [reply src/lib/symmetric.c]: ok, I wouldn't guess that. Can you please add comment why we do that, as it's not obvious?
I'll check if there is a way to improve botan's FFI api

Comment (ni4) [reply src/lib/symmetric.c]: Updated code with comment. Created issue #645 to sum up what we need to improve in botan

Comment (henrydcase) [reply src/lib/symmetric.c]: perfect

**Comment by ronaldtse** [comment]
Thank you @ni4 @dewyatt @henrydcase !
